{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BICS - Genome Analysis and Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"from google.colab import drive\\ndrive.mount('/content/drive')\\n\""
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Connect to Google Drive (make sure you have `metadata.tsv` in your Google Drive \"My Drive\" directory.)\n",
    "'''from google.colab import drive\n",
    "drive.mount('/content/drive')\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import defaultdict\n",
    "from tqdm import tqdm\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### metadata.tsv 733 MB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3063: DtypeWarning: Columns (13,16) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "#data = pd.read_csv(\"/content/drive/MyDrive/metadata.tsv\", sep=\"\\t\")\n",
    "data = pd.read_csv(\"./files/metadata.tsv\", sep=\"\\t\")\n",
    "\n",
    "#with open(\"/content/drive/MyDrive/reference.gb\") as f:\n",
    "with open(\"./files/reference.gb\") as f:\n",
    "  reference_full_string = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus name</th>\n",
       "      <th>Type</th>\n",
       "      <th>Accession ID</th>\n",
       "      <th>Collection date</th>\n",
       "      <th>Location</th>\n",
       "      <th>Additional location information</th>\n",
       "      <th>Sequence length</th>\n",
       "      <th>Host</th>\n",
       "      <th>Patient age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>...</th>\n",
       "      <th>Pangolin version</th>\n",
       "      <th>Variant</th>\n",
       "      <th>AA Substitutions</th>\n",
       "      <th>Submission date</th>\n",
       "      <th>Is reference?</th>\n",
       "      <th>Is complete?</th>\n",
       "      <th>Is high coverage?</th>\n",
       "      <th>Is low coverage?</th>\n",
       "      <th>N-Content</th>\n",
       "      <th>GC-Content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCoV-19/Australia/NT12/2020</td>\n",
       "      <td>betacoronavirus</td>\n",
       "      <td>EPI_ISL_426900</td>\n",
       "      <td>2020</td>\n",
       "      <td>Oceania / Australia / Northern territory</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29862</td>\n",
       "      <td>Human</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(NSP15_A283V,NSP12_P323L,Spike_D614G)</td>\n",
       "      <td>2020-04-17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.006912</td>\n",
       "      <td>0.379674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCoV-19/Australia/NT13/2020</td>\n",
       "      <td>betacoronavirus</td>\n",
       "      <td>EPI_ISL_426901</td>\n",
       "      <td>2020</td>\n",
       "      <td>Oceania / Australia / Northern territory</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29865</td>\n",
       "      <td>Human</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(NSP15_A283V,NSP12_P323L,Spike_D614G)</td>\n",
       "      <td>2020-04-17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.008305</td>\n",
       "      <td>0.379554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCoV-19/Australia/NT14/2020</td>\n",
       "      <td>betacoronavirus</td>\n",
       "      <td>EPI_ISL_426902</td>\n",
       "      <td>2020</td>\n",
       "      <td>Oceania / Australia / Northern territory</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29864</td>\n",
       "      <td>Human</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(NSP14_R163C,NSP3_K38R,NS3_G251V,NSP2_I559V,NS...</td>\n",
       "      <td>2020-04-17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.017929</td>\n",
       "      <td>0.378860</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Virus name             Type    Accession ID  \\\n",
       "0  hCoV-19/Australia/NT12/2020  betacoronavirus  EPI_ISL_426900   \n",
       "1  hCoV-19/Australia/NT13/2020  betacoronavirus  EPI_ISL_426901   \n",
       "2  hCoV-19/Australia/NT14/2020  betacoronavirus  EPI_ISL_426902   \n",
       "\n",
       "  Collection date                                  Location  \\\n",
       "0            2020  Oceania / Australia / Northern territory   \n",
       "1            2020  Oceania / Australia / Northern territory   \n",
       "2            2020  Oceania / Australia / Northern territory   \n",
       "\n",
       "  Additional location information  Sequence length   Host Patient age  \\\n",
       "0                             NaN            29862  Human     unknown   \n",
       "1                             NaN            29865  Human     unknown   \n",
       "2                             NaN            29864  Human     unknown   \n",
       "\n",
       "    Gender  ... Pangolin version Variant  \\\n",
       "0  unknown  ...       2021-05-19     NaN   \n",
       "1  unknown  ...       2021-05-19     NaN   \n",
       "2  unknown  ...       2021-05-19     NaN   \n",
       "\n",
       "                                    AA Substitutions Submission date  \\\n",
       "0              (NSP15_A283V,NSP12_P323L,Spike_D614G)      2020-04-17   \n",
       "1              (NSP15_A283V,NSP12_P323L,Spike_D614G)      2020-04-17   \n",
       "2  (NSP14_R163C,NSP3_K38R,NS3_G251V,NSP2_I559V,NS...      2020-04-17   \n",
       "\n",
       "  Is reference? Is complete? Is high coverage? Is low coverage? N-Content  \\\n",
       "0           NaN         True              True              NaN  0.006912   \n",
       "1           NaN         True              True              NaN  0.008305   \n",
       "2           NaN         True               NaN              NaN  0.017929   \n",
       "\n",
       "  GC-Content  \n",
       "0   0.379674  \n",
       "1   0.379554  \n",
       "2   0.378860  \n",
       "\n",
       "[3 rows x 22 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Preview data\n",
    "data.head(n=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1755981"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence length</th>\n",
       "      <th>N-Content</th>\n",
       "      <th>GC-Content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1.755981e+06</td>\n",
       "      <td>876815.000000</td>\n",
       "      <td>1.755981e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>2.931579e+04</td>\n",
       "      <td>0.033518</td>\n",
       "      <td>3.795408e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>3.530986e+03</td>\n",
       "      <td>0.974207</td>\n",
       "      <td>2.624725e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>6.400000e+01</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>1.538462e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>2.976300e+04</td>\n",
       "      <td>0.000738</td>\n",
       "      <td>3.795925e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>2.978200e+04</td>\n",
       "      <td>0.008552</td>\n",
       "      <td>3.797707e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>2.982800e+04</td>\n",
       "      <td>0.028928</td>\n",
       "      <td>3.799121e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>3.469200e+04</td>\n",
       "      <td>396.951613</td>\n",
       "      <td>5.131579e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sequence length      N-Content    GC-Content\n",
       "count     1.755981e+06  876815.000000  1.755981e+06\n",
       "mean      2.931579e+04       0.033518  3.795408e-01\n",
       "std       3.530986e+03       0.974207  2.624725e-03\n",
       "min       6.400000e+01       0.000033  1.538462e-01\n",
       "25%       2.976300e+04       0.000738  3.795925e-01\n",
       "50%       2.978200e+04       0.008552  3.797707e-01\n",
       "75%       2.982800e+04       0.028928  3.799121e-01\n",
       "max       3.469200e+04     396.951613  5.131579e-01"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_data = data\n",
    "data = full_data[::4]  # quarter of the entire dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reorganize Into Substitution Table\n",
    "## Make a table (dataframe), where each row represents a single AA substitution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "column_dict = {\n",
    "    \"name\": \"Virus name\",\n",
    "    \"date\": \"Collection date\",\n",
    "    \"age\": \"Patient age\",\n",
    "    \"gender\": \"Gender\",\n",
    "    \"version\": \"Pangolin version\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_substitutions(subs):\n",
    "  \"\"\"\n",
    "  Arguments\n",
    "  - subs: \"(NSP15_A283V,NSP12_P323L,Spike_D614G)\"\n",
    "  Returns\n",
    "  - List of individual substitutions: [\n",
    "    (\"NSP15_A283V\", \"NSP15\", \"A\", \"V\", 283),\n",
    "    ...\n",
    "  ]\n",
    "  \"\"\"\n",
    "  if pd.isna(subs):  # nan value\n",
    "    return []\n",
    "\n",
    "  subs = subs.replace(\"(\", \"\").replace(\")\", \"\")\n",
    "  subs = subs.split(\",\")\n",
    "  result = []\n",
    "  for s in subs:\n",
    "    try:\n",
    "      feature, mut = s.split(\"_\")\n",
    "      before = mut[0]\n",
    "      after = mut[-1]\n",
    "      index = int(mut[1:-1])\n",
    "    except:  # other types of substitutions\n",
    "      continue\n",
    "    result.append((s, feature, before, after, index))\n",
    "  return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████| 438996/438996 [03:26<00:00, 2130.40it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>date</th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>version</th>\n",
       "      <th>sub</th>\n",
       "      <th>sub_feature</th>\n",
       "      <th>sub_before</th>\n",
       "      <th>sub_after</th>\n",
       "      <th>sub_index</th>\n",
       "      <th>loc_continent</th>\n",
       "      <th>loc_country</th>\n",
       "      <th>loc_territory</th>\n",
       "      <th>loc_city</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCoV-19/Australia/NT12/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>NSP15_A283V</td>\n",
       "      <td>NSP15</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>283</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCoV-19/Australia/NT12/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>NSP12_P323L</td>\n",
       "      <td>NSP12</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>323</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCoV-19/Australia/NT12/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>Spike_D614G</td>\n",
       "      <td>Spike</td>\n",
       "      <td>D</td>\n",
       "      <td>G</td>\n",
       "      <td>614</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCoV-19/Australia/NT17/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>NSP12_P323L</td>\n",
       "      <td>NSP12</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>323</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hCoV-19/Australia/NT17/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>Spike_D614G</td>\n",
       "      <td>Spike</td>\n",
       "      <td>D</td>\n",
       "      <td>G</td>\n",
       "      <td>614</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          name  date      age   gender     version  \\\n",
       "0  hCoV-19/Australia/NT12/2020  2020  unknown  unknown  2021-05-19   \n",
       "1  hCoV-19/Australia/NT12/2020  2020  unknown  unknown  2021-05-19   \n",
       "2  hCoV-19/Australia/NT12/2020  2020  unknown  unknown  2021-05-19   \n",
       "3  hCoV-19/Australia/NT17/2020  2020  unknown  unknown  2021-05-19   \n",
       "4  hCoV-19/Australia/NT17/2020  2020  unknown  unknown  2021-05-19   \n",
       "\n",
       "           sub sub_feature sub_before sub_after  sub_index loc_continent  \\\n",
       "0  NSP15_A283V       NSP15          A         V        283       Oceania   \n",
       "1  NSP12_P323L       NSP12          P         L        323       Oceania   \n",
       "2  Spike_D614G       Spike          D         G        614       Oceania   \n",
       "3  NSP12_P323L       NSP12          P         L        323       Oceania   \n",
       "4  Spike_D614G       Spike          D         G        614       Oceania   \n",
       "\n",
       "  loc_country       loc_territory loc_city  \n",
       "0   Australia  Northern territory     None  \n",
       "1   Australia  Northern territory     None  \n",
       "2   Australia  Northern territory     None  \n",
       "3   Australia  Northern territory     None  \n",
       "4   Australia  Northern territory     None  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_data = defaultdict(list)\n",
    "for i, row in tqdm(data.iterrows(), total=len(data)):\n",
    "  # Identify mutations\n",
    "  subs = parse_substitutions(row[\"AA Substitutions\"])\n",
    "  location_info = [s.strip() for s in row[\"Location\"].split(\"/\")]\n",
    "  continent = location_info[0]\n",
    "  country = location_info[1]\n",
    "  city = None\n",
    "  territory = None\n",
    "  if len(location_info) > 2:\n",
    "    territory = location_info[2]\n",
    "    if len(location_info) > 3:\n",
    "      city = location_info[3]\n",
    "\n",
    "  # Copy over columns from the column dict\n",
    "  for new_column, old_column in column_dict.items():\n",
    "    sub_data[new_column].extend([row[old_column]] * len(subs))\n",
    "\n",
    "  for full_sub, feature, before, after, index in subs:\n",
    "    # Fill mutation columns\n",
    "    sub_data[\"sub\"].append(full_sub)\n",
    "    sub_data[\"sub_feature\"].append(feature)\n",
    "    sub_data[\"sub_before\"].append(before)\n",
    "    sub_data[\"sub_after\"].append(after)\n",
    "    sub_data[\"sub_index\"].append(index)\n",
    "\n",
    "    # Fill location columns\n",
    "    sub_data[\"loc_continent\"].append(continent)\n",
    "    sub_data[\"loc_country\"].append(country)\n",
    "    sub_data[\"loc_territory\"].append(territory)\n",
    "    sub_data[\"loc_city\"].append(city)\n",
    "\n",
    "sub_df = pd.DataFrame(sub_data)\n",
    "sub_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "438996"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6557926"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sub_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([None, 'Sydney', 'Rotterdam', ..., 'Lot', 'Tarn-et-Garonne', 'Gers'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_df.loc_city.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    NSP15_A283V\n",
       "1    NSP12_P323L\n",
       "2    Spike_D614G\n",
       "3    NSP12_P323L\n",
       "4    Spike_D614G\n",
       "Name: sub, dtype: object"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_df[\"sub\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['NSP15_A283V',\n",
       " 'NSP9_A108T',\n",
       " 'NSP9_D50B',\n",
       " 'NSP6_T20A',\n",
       " 'NS3_I7F',\n",
       " 'NSP6_E250D',\n",
       " 'N_D415V',\n",
       " 'NSP3_G1423D',\n",
       " 'NSP3_E1733K']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_subs = sub_df[\"sub\"].unique().tolist()\n",
    "unique_subs[::5000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40592"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(unique_subs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Spike_D614G    427632\n",
       "NSP12_P323L    416431\n",
       "N_R203K        247788\n",
       "N_G204R        233851\n",
       "Spike_N501Y    201407\n",
       "                ...  \n",
       "NSP12_A125T         1\n",
       "NSP15_G156C         1\n",
       "Spike_N928D         1\n",
       "NS8_P85Q            1\n",
       "NSP12_A376C         1\n",
       "Name: sub, Length: 40592, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_df[\"sub\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['E', 'M', 'N', 'NS3', 'NS6', 'NS7a', 'NS7b', 'NS8', 'NSP1', 'NSP10', 'NSP12', 'NSP13', 'NSP14', 'NSP15', 'NSP16', 'NSP2', 'NSP3', 'NSP4', 'NSP5', 'NSP6', 'NSP7', 'NSP8', 'NSP9', 'Spike']\n"
     ]
    }
   ],
   "source": [
    "print(sorted(sub_df[\"sub_feature\"].unique().tolist()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation Index\n",
    "Let's figure out where each mutation occured, within the full genome sequence. Let's look at this key: NSP12_D291N.\n",
    "\n",
    "This means that the mutation occured in NSP12 and the exact location is index 291 (relative to the start of NSP12).\n",
    "\n",
    "We can use this information to find the exact location of each substituion."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find Feature Locations From GenBank File\n",
    "We need to find the start index of each feature (including proteins, peptides, etc.)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_feature_locations(genbank_path):\n",
    "  with open(genbank_path) as f:\n",
    "    gb = f.read()\n",
    "  gb_oneline = gb.replace(\"\\n\", \"\")\n",
    "  gen_regex = r'gene.+?(\\d*)\\.\\.(\\d*).+?\\/gene=\"(.+?)\"'\n",
    "  pep_regex = r'mat_peptide.+?(\\d*)\\.\\.(\\d*).+?/gene=\"(.+?)\".+?/product=\"(.+?)\".+?/note=\"(.{10})'\n",
    "  gen_data = re.findall(gen_regex, gb_oneline)\n",
    "  pep_data = re.findall(pep_regex, gb_oneline)\n",
    "\n",
    "  # Gene Ranges\n",
    "  range_dict = dict()\n",
    "  letters = [\"E\", \"M\", \"N\", \"S\"]\n",
    "  for start, end, name in gen_data:\n",
    "      start = int(start)\n",
    "      for letter in letters:\n",
    "          if letter in name:\n",
    "              range_dict[name] = (start - 1, end)\n",
    "\n",
    "  # Peptide Ranges\n",
    "  candidates = []\n",
    "  for i in range(1, 17):\n",
    "      candidates.append(\"nsp{}\".format(i))\n",
    "  candidates.reverse()  # e.g., check nsp11 before nsp1\n",
    "  for start, end, _, name, note in pep_data:\n",
    "      start = int(start)\n",
    "      end = int(end)\n",
    "      for c in candidates:\n",
    "          if c in name.lower() or c in note.lower():\n",
    "              range_dict[c.upper()] = (start - 1, end)\n",
    "              \n",
    "  # Spike Protein Ranges\n",
    "  range_dict[\"Spike\"] = range_dict[\"S\"]\n",
    "  del range_dict[\"S\"]\n",
    "\n",
    "  return dict(range_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'E': (26244, '26472'), 'M': (26522, '27191'), 'N': (28273, '29533'), 'NSP1': (13441, 13480), 'NSP2': (805, 2719), 'NSP3': (2719, 8554), 'NSP4': (8554, 10054), 'NSP5': (10054, 10972), 'NSP6': (10972, 11842), 'NSP7': (11842, 12091), 'NSP8': (12091, 12685), 'NSP9': (12685, 13024), 'NSP10': (13024, 13441), 'NSP12': (13441, 13468), 'NSP13': (16236, 18039), 'NSP14': (18039, 19620), 'NSP15': (19620, 20658), 'NSP16': (20658, 21552), 'NSP11': (13441, 13480), 'Spike': (21562, '25384')}\n"
     ]
    }
   ],
   "source": [
    "#range_dict = parse_feature_locations(\"/content/drive/MyDrive/reference.gb\")\n",
    "range_dict = parse_feature_locations(\"./files/reference.gb\")\n",
    "\n",
    "print(range_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter Features\n",
    "### Filter rows that contain features for which we have the starting index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>date</th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>version</th>\n",
       "      <th>sub</th>\n",
       "      <th>sub_feature</th>\n",
       "      <th>sub_before</th>\n",
       "      <th>sub_after</th>\n",
       "      <th>sub_index</th>\n",
       "      <th>loc_continent</th>\n",
       "      <th>loc_country</th>\n",
       "      <th>loc_territory</th>\n",
       "      <th>loc_city</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCoV-19/Australia/NT12/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>NSP15_A283V</td>\n",
       "      <td>NSP15</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>283</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCoV-19/Australia/NT12/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>NSP12_P323L</td>\n",
       "      <td>NSP12</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>323</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCoV-19/Australia/NT12/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>Spike_D614G</td>\n",
       "      <td>Spike</td>\n",
       "      <td>D</td>\n",
       "      <td>G</td>\n",
       "      <td>614</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCoV-19/Australia/NT17/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>NSP12_P323L</td>\n",
       "      <td>NSP12</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>323</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hCoV-19/Australia/NT17/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>Spike_D614G</td>\n",
       "      <td>Spike</td>\n",
       "      <td>D</td>\n",
       "      <td>G</td>\n",
       "      <td>614</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6557921</th>\n",
       "      <td>hCoV-19/Croatia/1692/2021</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>70</td>\n",
       "      <td>Male</td>\n",
       "      <td>2021-05-27</td>\n",
       "      <td>NSP12_P323L</td>\n",
       "      <td>NSP12</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>323</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Croatia</td>\n",
       "      <td>Požega Slavonia County</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6557922</th>\n",
       "      <td>hCoV-19/Croatia/1692/2021</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>70</td>\n",
       "      <td>Male</td>\n",
       "      <td>2021-05-27</td>\n",
       "      <td>Spike_D614G</td>\n",
       "      <td>Spike</td>\n",
       "      <td>D</td>\n",
       "      <td>G</td>\n",
       "      <td>614</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Croatia</td>\n",
       "      <td>Požega Slavonia County</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6557923</th>\n",
       "      <td>hCoV-19/Croatia/1692/2021</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>70</td>\n",
       "      <td>Male</td>\n",
       "      <td>2021-05-27</td>\n",
       "      <td>N_D3L</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "      <td>L</td>\n",
       "      <td>3</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Croatia</td>\n",
       "      <td>Požega Slavonia County</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6557924</th>\n",
       "      <td>hCoV-19/Croatia/1692/2021</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>70</td>\n",
       "      <td>Male</td>\n",
       "      <td>2021-05-27</td>\n",
       "      <td>Spike_S982A</td>\n",
       "      <td>Spike</td>\n",
       "      <td>S</td>\n",
       "      <td>A</td>\n",
       "      <td>982</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Croatia</td>\n",
       "      <td>Požega Slavonia County</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6557925</th>\n",
       "      <td>hCoV-19/Croatia/1692/2021</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>70</td>\n",
       "      <td>Male</td>\n",
       "      <td>2021-05-27</td>\n",
       "      <td>N_S235F</td>\n",
       "      <td>N</td>\n",
       "      <td>S</td>\n",
       "      <td>F</td>\n",
       "      <td>235</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Croatia</td>\n",
       "      <td>Požega Slavonia County</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5722571 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                name        date      age   gender  \\\n",
       "0        hCoV-19/Australia/NT12/2020        2020  unknown  unknown   \n",
       "1        hCoV-19/Australia/NT12/2020        2020  unknown  unknown   \n",
       "2        hCoV-19/Australia/NT12/2020        2020  unknown  unknown   \n",
       "3        hCoV-19/Australia/NT17/2020        2020  unknown  unknown   \n",
       "4        hCoV-19/Australia/NT17/2020        2020  unknown  unknown   \n",
       "...                              ...         ...      ...      ...   \n",
       "6557921    hCoV-19/Croatia/1692/2021  2021-03-12       70     Male   \n",
       "6557922    hCoV-19/Croatia/1692/2021  2021-03-12       70     Male   \n",
       "6557923    hCoV-19/Croatia/1692/2021  2021-03-12       70     Male   \n",
       "6557924    hCoV-19/Croatia/1692/2021  2021-03-12       70     Male   \n",
       "6557925    hCoV-19/Croatia/1692/2021  2021-03-12       70     Male   \n",
       "\n",
       "            version          sub sub_feature sub_before sub_after  sub_index  \\\n",
       "0        2021-05-19  NSP15_A283V       NSP15          A         V        283   \n",
       "1        2021-05-19  NSP12_P323L       NSP12          P         L        323   \n",
       "2        2021-05-19  Spike_D614G       Spike          D         G        614   \n",
       "3        2021-05-19  NSP12_P323L       NSP12          P         L        323   \n",
       "4        2021-05-19  Spike_D614G       Spike          D         G        614   \n",
       "...             ...          ...         ...        ...       ...        ...   \n",
       "6557921  2021-05-27  NSP12_P323L       NSP12          P         L        323   \n",
       "6557922  2021-05-27  Spike_D614G       Spike          D         G        614   \n",
       "6557923  2021-05-27        N_D3L           N          D         L          3   \n",
       "6557924  2021-05-27  Spike_S982A       Spike          S         A        982   \n",
       "6557925  2021-05-27      N_S235F           N          S         F        235   \n",
       "\n",
       "        loc_continent loc_country           loc_territory loc_city  \n",
       "0             Oceania   Australia      Northern territory     None  \n",
       "1             Oceania   Australia      Northern territory     None  \n",
       "2             Oceania   Australia      Northern territory     None  \n",
       "3             Oceania   Australia      Northern territory     None  \n",
       "4             Oceania   Australia      Northern territory     None  \n",
       "...               ...         ...                     ...      ...  \n",
       "6557921        Europe     Croatia  Požega Slavonia County     None  \n",
       "6557922        Europe     Croatia  Požega Slavonia County     None  \n",
       "6557923        Europe     Croatia  Požega Slavonia County     None  \n",
       "6557924        Europe     Croatia  Požega Slavonia County     None  \n",
       "6557925        Europe     Croatia  Požega Slavonia County     None  \n",
       "\n",
       "[5722571 rows x 14 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_sub_df = sub_df\n",
    "\n",
    "sub_df = sub_df[sub_df[\"sub_feature\"].isin(range_dict.keys())]\n",
    "sub_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add Mutation Index To Table\n",
    "Add the following columns based on the ranges obtained above:\n",
    "\n",
    "* sub_feature_start\n",
    "* sub_full_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\ipykernel_launcher.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  after removing the cwd from sys.path.\n",
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\ipykernel_launcher.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \"\"\"\n"
     ]
    }
   ],
   "source": [
    "start_dict = {key: value[0] for key, value in range_dict.items()}\n",
    "map_start_index = (lambda feature: start_dict[feature])\n",
    "\n",
    "sub_df[\"sub_feature_start\"] = sub_df[\"sub_feature\"].apply(map_start_index)\n",
    "sub_df[\"sub_full_index\"] = sub_df[\"sub_feature_start\"] + sub_df[\"sub_index\"] * 3\n",
    "# assumes that sub index is the AA index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>date</th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>version</th>\n",
       "      <th>sub</th>\n",
       "      <th>sub_feature</th>\n",
       "      <th>sub_before</th>\n",
       "      <th>sub_after</th>\n",
       "      <th>sub_index</th>\n",
       "      <th>loc_continent</th>\n",
       "      <th>loc_country</th>\n",
       "      <th>loc_territory</th>\n",
       "      <th>loc_city</th>\n",
       "      <th>sub_feature_start</th>\n",
       "      <th>sub_full_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCoV-19/Australia/NT12/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>NSP15_A283V</td>\n",
       "      <td>NSP15</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>283</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "      <td>19620</td>\n",
       "      <td>20469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCoV-19/Australia/NT12/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>NSP12_P323L</td>\n",
       "      <td>NSP12</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>323</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "      <td>13441</td>\n",
       "      <td>14410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCoV-19/Australia/NT12/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>Spike_D614G</td>\n",
       "      <td>Spike</td>\n",
       "      <td>D</td>\n",
       "      <td>G</td>\n",
       "      <td>614</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "      <td>21562</td>\n",
       "      <td>23404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCoV-19/Australia/NT17/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>NSP12_P323L</td>\n",
       "      <td>NSP12</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>323</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "      <td>13441</td>\n",
       "      <td>14410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hCoV-19/Australia/NT17/2020</td>\n",
       "      <td>2020</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2021-05-19</td>\n",
       "      <td>Spike_D614G</td>\n",
       "      <td>Spike</td>\n",
       "      <td>D</td>\n",
       "      <td>G</td>\n",
       "      <td>614</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>None</td>\n",
       "      <td>21562</td>\n",
       "      <td>23404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6557921</th>\n",
       "      <td>hCoV-19/Croatia/1692/2021</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>70</td>\n",
       "      <td>Male</td>\n",
       "      <td>2021-05-27</td>\n",
       "      <td>NSP12_P323L</td>\n",
       "      <td>NSP12</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>323</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Croatia</td>\n",
       "      <td>Požega Slavonia County</td>\n",
       "      <td>None</td>\n",
       "      <td>13441</td>\n",
       "      <td>14410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6557922</th>\n",
       "      <td>hCoV-19/Croatia/1692/2021</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>70</td>\n",
       "      <td>Male</td>\n",
       "      <td>2021-05-27</td>\n",
       "      <td>Spike_D614G</td>\n",
       "      <td>Spike</td>\n",
       "      <td>D</td>\n",
       "      <td>G</td>\n",
       "      <td>614</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Croatia</td>\n",
       "      <td>Požega Slavonia County</td>\n",
       "      <td>None</td>\n",
       "      <td>21562</td>\n",
       "      <td>23404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6557923</th>\n",
       "      <td>hCoV-19/Croatia/1692/2021</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>70</td>\n",
       "      <td>Male</td>\n",
       "      <td>2021-05-27</td>\n",
       "      <td>N_D3L</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "      <td>L</td>\n",
       "      <td>3</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Croatia</td>\n",
       "      <td>Požega Slavonia County</td>\n",
       "      <td>None</td>\n",
       "      <td>28273</td>\n",
       "      <td>28282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6557924</th>\n",
       "      <td>hCoV-19/Croatia/1692/2021</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>70</td>\n",
       "      <td>Male</td>\n",
       "      <td>2021-05-27</td>\n",
       "      <td>Spike_S982A</td>\n",
       "      <td>Spike</td>\n",
       "      <td>S</td>\n",
       "      <td>A</td>\n",
       "      <td>982</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Croatia</td>\n",
       "      <td>Požega Slavonia County</td>\n",
       "      <td>None</td>\n",
       "      <td>21562</td>\n",
       "      <td>24508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6557925</th>\n",
       "      <td>hCoV-19/Croatia/1692/2021</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>70</td>\n",
       "      <td>Male</td>\n",
       "      <td>2021-05-27</td>\n",
       "      <td>N_S235F</td>\n",
       "      <td>N</td>\n",
       "      <td>S</td>\n",
       "      <td>F</td>\n",
       "      <td>235</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Croatia</td>\n",
       "      <td>Požega Slavonia County</td>\n",
       "      <td>None</td>\n",
       "      <td>28273</td>\n",
       "      <td>28978</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5722571 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                name        date      age   gender  \\\n",
       "0        hCoV-19/Australia/NT12/2020        2020  unknown  unknown   \n",
       "1        hCoV-19/Australia/NT12/2020        2020  unknown  unknown   \n",
       "2        hCoV-19/Australia/NT12/2020        2020  unknown  unknown   \n",
       "3        hCoV-19/Australia/NT17/2020        2020  unknown  unknown   \n",
       "4        hCoV-19/Australia/NT17/2020        2020  unknown  unknown   \n",
       "...                              ...         ...      ...      ...   \n",
       "6557921    hCoV-19/Croatia/1692/2021  2021-03-12       70     Male   \n",
       "6557922    hCoV-19/Croatia/1692/2021  2021-03-12       70     Male   \n",
       "6557923    hCoV-19/Croatia/1692/2021  2021-03-12       70     Male   \n",
       "6557924    hCoV-19/Croatia/1692/2021  2021-03-12       70     Male   \n",
       "6557925    hCoV-19/Croatia/1692/2021  2021-03-12       70     Male   \n",
       "\n",
       "            version          sub sub_feature sub_before sub_after  sub_index  \\\n",
       "0        2021-05-19  NSP15_A283V       NSP15          A         V        283   \n",
       "1        2021-05-19  NSP12_P323L       NSP12          P         L        323   \n",
       "2        2021-05-19  Spike_D614G       Spike          D         G        614   \n",
       "3        2021-05-19  NSP12_P323L       NSP12          P         L        323   \n",
       "4        2021-05-19  Spike_D614G       Spike          D         G        614   \n",
       "...             ...          ...         ...        ...       ...        ...   \n",
       "6557921  2021-05-27  NSP12_P323L       NSP12          P         L        323   \n",
       "6557922  2021-05-27  Spike_D614G       Spike          D         G        614   \n",
       "6557923  2021-05-27        N_D3L           N          D         L          3   \n",
       "6557924  2021-05-27  Spike_S982A       Spike          S         A        982   \n",
       "6557925  2021-05-27      N_S235F           N          S         F        235   \n",
       "\n",
       "        loc_continent loc_country           loc_territory loc_city  \\\n",
       "0             Oceania   Australia      Northern territory     None   \n",
       "1             Oceania   Australia      Northern territory     None   \n",
       "2             Oceania   Australia      Northern territory     None   \n",
       "3             Oceania   Australia      Northern territory     None   \n",
       "4             Oceania   Australia      Northern territory     None   \n",
       "...               ...         ...                     ...      ...   \n",
       "6557921        Europe     Croatia  Požega Slavonia County     None   \n",
       "6557922        Europe     Croatia  Požega Slavonia County     None   \n",
       "6557923        Europe     Croatia  Požega Slavonia County     None   \n",
       "6557924        Europe     Croatia  Požega Slavonia County     None   \n",
       "6557925        Europe     Croatia  Požega Slavonia County     None   \n",
       "\n",
       "         sub_feature_start  sub_full_index  \n",
       "0                    19620           20469  \n",
       "1                    13441           14410  \n",
       "2                    21562           23404  \n",
       "3                    13441           14410  \n",
       "4                    21562           23404  \n",
       "...                    ...             ...  \n",
       "6557921              13441           14410  \n",
       "6557922              21562           23404  \n",
       "6557923              28273           28282  \n",
       "6557924              21562           24508  \n",
       "6557925              28273           28978  \n",
       "\n",
       "[5722571 rows x 16 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Congrats!\n",
    "We've successfully pre-processed the data into the substitution table. This should help us analyze and visualize stats, related to gene substitutions, much more easily."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
